paddle/fluid/framework/parallel_executor.cc

```diff
@@ -0,0 +1,673 @@
+    std::stringstream ss;
+    ss << name_ << ":" << place_;
+    return ss.str();
+  }
+
+  size_t version_;
+  std::string name_;
+  platform::Place place_;
+};
+
+struct DependencyVarHandle : public VarHandleBase {
+  std::string DebugString() const override { return "Dependency Variable"; }
+};
+
+struct OpHandle {
+  std::vector<VarHandleBase *> inputs_;
+  std::vector<VarHandleBase *> outputs_;
+  std::unordered_map<platform::Place, platform::DeviceContext *,
+                     platform::PlaceHash>
+      dev_ctx_;
+
+  std::string DebugString() {
+    std::stringstream ss;
+    ss << "(";
+    for (auto *var : inputs_) {
+      ss << var->DebugString() << ", ";
+    }
+    ss << ") --> (";
+    for (auto *var : outputs_) {
+      ss << var->DebugString() << ", ";
+    }
+    ss << ")\n";
+    return ss.str();
+  }
+
+  virtual ~OpHandle() {}
+
+  virtual void Run() { PADDLE_THROW("Not implemented"); }
+  virtual void Wait(platform::DeviceContext *waited_dev) {}
+};
+
+struct ComputationOpHandle : public OpHandle {
+  std::unique_ptr<OperatorBase> op_;
+  Scope *scope_;
+  platform::Place place_;
+
+  explicit ComputationOpHandle(const OpDesc &op_desc, Scope *scope,
+                               platform::Place place)
+      : op_(framework::OpRegistry::CreateOp(op_desc)),
+        scope_(scope),
+        place_(place) {}
+
+  void Run() override {
+    // Wait other op if necessary
+    LOG(INFO) << "Run " << this << " " << DebugString();
+    auto *cur_ctx = dev_ctx_[place_];
+    for (auto *in : inputs_) {
+      if (in->generated_op_ && in->generated_op_->dev_ctx_[place_] != cur_ctx) {
+        in->generated_op_->Wait(cur_ctx);
+      }
+    }
+
+    op_->Run(*scope_, place_);
+    LOG(INFO) << "Done " << this;
+  }
+
+  void Wait(platform::DeviceContext *waited_dev) override {
+    this->dev_ctx_.at(place_)->Wait();
+  }
+};
+
+struct ScaleLossGradOpHandle : public OpHandle {
+  float coeff_;
+  Scope *scope_;
+  platform::Place place_;
+
+  explicit ScaleLossGradOpHandle(size_t num_dev, Scope *scope,
+                                 platform::Place place)
+      : coeff_(static_cast<float>(1.0 / num_dev)),
+        scope_(scope),
+        place_(place) {}
+
+  void Run() override {
+    LOG(INFO) << "Run Scale Loss Grad";
+
+    std::string var_name = static_cast<VarHandle *>(this->outputs_[0])->name_;
+
+    float *tmp = scope_->FindVar(var_name)
+                     ->GetMutable<framework::LoDTensor>()
+                     ->mutable_data<float>(make_ddim({1}), place_);
+
+    if (platform::is_cpu_place(place_)) {
+      *tmp = coeff_;
+    } else {
+      memory::Copy(
+          boost::get<platform::CUDAPlace>(place_), tmp, platform::CPUPlace(),
+          &coeff_, sizeof(float),
+          static_cast<platform::CUDADeviceContext *>(this->dev_ctx_[place_])
+              ->stream());
+    }
+  }
+
+  void Wait(platform::DeviceContext *waited_dev) override {
+    this->dev_ctx_.at(place_)->Wait();
+  }
```

Comment:
`waited_dev` is not used.
`Wait` interface waits on which cuda stream? Its own or passed parameter?